mockgcp/mockstorage/v1.go

```diff
@@ -0,0 +1,427 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mockstorage
+
+import (
+	"context"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	pb "github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/generated/mockgcp/storage/v1"
+	"github.com/golang/protobuf/ptypes/empty"
+)
+
+type StorageV1 struct {
+	*MockService
+	pb.UnimplementedStorageServer
+}
+
+func (s *StorageV1) GetBucket(ctx context.Context, req *pb.GetBucketRequest) (*pb.Bucket, error) {
+	fqn := req.GetBucket()
+
+	now := timestamppb.Now()
+
+	obj := &pb.Bucket{}
+	obj.Id = fqn
+	obj.Name = fqn
+	obj.ProjectNumber = 0 // todo
+	obj.Location = "US"
+	obj.Location = "multi-region"
+	obj.StorageClass = "STANDARD"
+	obj.TimeCreated = now
+	obj.Updated = now
+
+	if err := s.storage.Get(ctx, fqn, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, status.Errorf(codes.NotFound, "bucket %q not found", fqn)
+		} else {
+			return nil, status.Errorf(codes.Internal, "error reading bucket: %v", err)
+		}
+	}
+
+	return obj, nil
+}
+
+func (s *StorageV1) InsertBucket(ctx context.Context, req *pb.InsertBucketRequest) (*pb.Bucket, error) {
+	fqn := req.GetBucket().GetName()
+
+	obj := proto.Clone(req.GetBucket()).(*pb.Bucket)
+	obj.Name = fqn
+
+	if err := s.storage.Create(ctx, fqn, obj); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			return nil, status.Errorf(codes.AlreadyExists, "bucket already exists")
+		}
+		return nil, status.Errorf(codes.Internal, "error creating bucket: %v", err)
+	}
+
+	return obj, nil
+}
+
+func (s *StorageV1) PatchBucket(ctx context.Context, req *pb.PatchBucketRequest) (*pb.Bucket, error) {
+	fqn := req.GetBucket()
+
+	obj := &pb.Bucket{}
+	if err := s.storage.Get(ctx, fqn, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, status.Errorf(codes.NotFound, "bucket %q not found", fqn)
+		}
+		return nil, status.Errorf(codes.Internal, "error reading bucket: %v", err)
+	}
+
+	// // Required. The update mask applies to the resource.
+	// paths := req.GetUpdateMask().GetPaths()
+	// if len(paths) == 0 {
+	// 	klog.Warningf("update_mask was not provided in request, should be required")
+	// }
+
+	// // TODO: Some sort of helper for fieldmask?
+	// for _, path := range paths {
+	// 	switch path {
+	// 	case "description":
+	// 		obj.Description = req.GetCertificate().GetDescription()
+	// 	case "labels":
+	// 		obj.Labels = req.GetCertificate().GetLabels()
+	// 	default:
+	// 		return nil, status.Errorf(codes.InvalidArgument, "update_mask path %q not valid", path)
+	// 	}
+	// }
```

Comment:
nit: remove commented lines

Reply:
Good catch - done (and there were more later on also!)